### PR TITLE
Humdrum import: Add @delayed="true" for delayed turns.

### PR DESCRIPTION
**Pull request by gregchapman-dev**

### CODE REVIEW

**Comment by craigsapp** [comment]
Turns are able to be delayed currently:

Screenshot 2023-01-13 at 3 38 52 PM

Humdrum test data:

```tsv
**kern
*M4/4
!LO:TX:b:t=not delayed
2dsSss
2e
=1
!LO:TX:b:t=delayed
2dSss
2e
==
*-
```

<details>
<summary> Click to view MEI conversion of above Humdrum data for example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-01-13T15:45:18" version="3.15.0-dev-72c755c">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="melbivv">
    <score xml:id="s7dzl2r">
     <scoreDef xml:id="s20f06h">
      <staffGrp xml:id="svyuwe1">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="c1rozvi4" shape="G" line="2" />
        <meterSig xml:id="metersig-L2F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="2" oct="4" pname="d" accid.ges="n" />
         <note xml:id="note-L5F1" dur="2" oct="4" pname="e">
          <accid xml:id="accid-L5F1" accid="n" func="caution" />
         </note>
        </layer>
       </staff>
       <turn xml:id="turn-L4F1" staff="1" startid="#note-L4F1" accidupper="f" accidlower="s" form="upper" />
       <dir xml:id="dir-L3F1" place="below" staff="1" tstamp="1.000000">
        <rend xml:id="rt1dpgs" fontstyle="normal">not delayed</rend>
       </dir>
      </measure>
      <measure xml:id="measure-L6" right="end" n="1">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L8F1" dur="2" oct="4" pname="d" accid.ges="n" />
         <note xml:id="note-L9F1" dur="2" oct="4" pname="e">
          <accid xml:id="accid-L9F1" accid="n" func="caution" />
         </note>
        </layer>
       </staff>
       <turn xml:id="turn-L8F1" staff="1" tstamp="2.000000" accidupper="f" accidlower="s" form="upper" />
       <dir xml:id="dir-L7F1" place="below" staff="1" tstamp="1.000000">
        <rend xml:id="rqdivzt" fontstyle="normal">delayed</rend>
       </dir>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Doing some tests, I see that using `@delayed="true"` is ignored when adding to the score using `@tstamp` to place the turn.

When `@startid` is instead used to link to a specific note, then `@delayed="true"` will delay the turn by 1/2 of the duration of the note.

I think it would be best to change from using `@tstamp` to `@startid`.   Then (1) `@delayed` will not be ignored, and (2) this is a more semantic translation of the Humdrum score into MEI.   

Do you want to also switch to `@startid` in this PR as well?

**Comment by gregchapman-dev** [comment]
How about this?  I will use @startid if possible, and a modified @tstamp otherwise.

Note that I have now also fixed another (vaguely) related bug, where mordents and trills will get a @startid that doesn't actually reference an existing note, because they used a chord id with a "note-" prefix.  This happens about 30x (to mordents) when importing mazurka30-4.krn (from the Chopin Mazurkas set of scores).  I have simply replaced the addMordent and and addTrill startid-setting code with the (correct) code from addTurn.

**Comment by craigsapp** [comment]
It would be good to give short examples highlighting the changes in the code as well as citations of entire compositions &mdash; with sample conversions before and after the fix so that I don't have to remember how to download a PR to compile and run on data tests myself 😛 .  Here is a study of mordents before the fix:

```tsv
**kern
*clefG2
*M4/4
=1
2cM
2d
=2
2c 2em
2d
=3
2cM 2em
2d
=4
2cMy 2em
2d
=
*-
```

<details>
<summary> Click to view MEI conversion of above Humdrum data </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-01-16T08:46:26" version="3.15.0-dev-244e188-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m1e2p8ey">
    <score xml:id="s1rv5qzy">
     <scoreDef xml:id="s144y9ud">
      <staffGrp xml:id="sarl2hm">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
        <meterSig xml:id="metersig-L3F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="2" oct="4" pname="c" accid.ges="n" />
         <note xml:id="note-L6F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <mordent xml:id="mordent-L5F1" staff="1" startid="#note-L5F1" place="above" form="upper" />
      </measure>
      <measure xml:id="measure-L7" n="2">
       <staff xml:id="staff-L7F1N1" n="1">
        <layer xml:id="layer-L7F1N1" n="1">
         <chord xml:id="chord-L8F1" dur="2">
          <note xml:id="note-L8F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L8F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L9F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <mordent xml:id="mordent-L8F1" staff="1" startid="#note-L8F1" place="above" form="upper" />
      </measure>
      <measure xml:id="measure-L10" n="3">
       <staff xml:id="staff-L10F1N1" n="1">
        <layer xml:id="layer-L10F1N1" n="1">
         <chord xml:id="chord-L11F1" dur="2">
          <note xml:id="note-L11F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L11F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L12F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <mordent xml:id="mordent-L11F1S1" staff="1" startid="#note-L11F1S1" place="below" form="upper" />
       <mordent xml:id="mordent-L11F1S2" staff="1" startid="#note-L11F1S2" place="above" form="upper" />
      </measure>
      <measure xml:id="measure-L13" n="4">
       <staff xml:id="staff-L13F1N1" n="1">
        <layer xml:id="layer-L13F1N1" n="1">
         <chord xml:id="chord-L14F1" dur="2">
          <note xml:id="note-L14F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L14F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L15F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <mordent xml:id="mordent-L14F1" staff="1" startid="#note-L14F1" place="above" form="upper" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>


Current conversion/rendering:

Screenshot 2023-01-16 at 8 36 56 AM

> because they used a chord id with a "note-" prefix.

Measures 2 and 4 are incorrectly rendered due to this problem.

For measure 2, the mordent is missing the chord note enumeration (`S2` meaning 2nd subtoken):

```xml
       <mordent xml:id="mordent-L8F1" staff="1" startid="#note-L8F1" place="above" form="upper" />
```

The mordent should be attached to the E4 note (not to the chord):

```xml
         <chord xml:id="chord-L8F1" dur="2">
          <note xml:id="note-L8F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L8F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
```

So the correct conversion for the mordent should be:

```xml
       <mordent xml:id="mordent-L8F1" staff="1" startid="#note-L8F1S2" place="above" form="upper" />
```

Is that what this PR does?  In other words, it should not be `mordent@startid="#chord-L8F1"` because usually a chord with a mordent on both notes will look like measure three (two separate mordents for the two separate notes).   When the mordent is attached to the chord, verovio only renders one of them, which is not the typical rendering style for mordents (but that is a reasonable assumption for what verovio should do). 

The standard interpretation of a mordent on a chord is that only the top note of the chord has the mordent applied to it.   This is unlike articulations such as staccatos which applies to all notes in the chords.  So attaching a mordent to a chord is problematic for how to interpret it.

Measure 4 would be a case where both notes have mordents, but only one mordent should be shown.  For practical purposes this is converted the same way as measure 2, where the mordent on the bottom note is not converted because there is no `mordent@visible="false"` in MEI, and as mentioned above, a mordent at the chord level is ambiguous and would otherwise imply that the bottom note should not be mordented.   Using `mordent@sameas` is debatable to be used here: visually it would be true, but for gestural purposes that would be strange (the mordents have different pitches, and also different intervals of a major second on the bottom and minor second on the top).

So to summarize, mordents do not operate on the chord level in `**kern` data, and should not as well in the MEI conversion.  This is different from articulations, which should operate on the note level in `**kern` data, but are allowed at the chord level in MEI:

Trills will behave in a similar manner to mordents.  I will make some test examples for them in the next post (and also ask if the new code fixes the IDs to point to the chord note, which is preferred, rather than the chord).

**Comment by gregchapman-dev** [comment]
My current fix to mordent and trill follows the turn code, which attaches the turn to the chord (not the note within the chord).

Sounds like I need to fix mordent and trill differently, to reference the note within the chord, and then fix turn as well, to do the same.

Thanks for the simple example, I’ll try to do better next time.

**Comment by craigsapp** [comment]
> Sounds like I need to fix mordent and trill differently, to reference the note within the chord, and then fix turn as well, to do the same.

Yes, turns are associated to a specific note in a chord as well (but note that fermatas are not, since they may also apply to note/chords in other voices/layers on the staff as well as to the note/chord to which they seem to be attached (so it is better to use `@tstamp` for fermatas, although they will have to be attached to notes in `**mens` (mensural music) conversions to MEI since `@tstamp` has a nearly useless meaning in MEI-mensural due to lack of `<measures>` and mensural rhythms being more like pitch intervals rather than absolute pitch since the performance rhythm contextual to "alterations" and such.).

> Thanks for the simple example, I’ll try to do better next time.

No problem as these are fine points.  😃 

Trills will behave similar to mordents/turns in the Humdrum-to-MEI converter: they should always attach to a note, and if there is a trill in a chord, still attach to the specific note(s).  

In piano music this is rare since they are hard to play, and usually occur only on thirds.   Here is an example from Beethoven's 4th piano concerto:

https://imslp.org/wiki/Piano_Concerto_No.4%2C_Op.58_(Beethoven%2C_Ludwig_van)
http://vmirror.imslp.org/files/imglnks/usimg/7/7b/IMSLP01287-Piano_Concerto_No._4_in_G_Major,_Op._58-I._Allegro_moderato.pdf

Screenshot 2023-01-16 at 12 16 41 PM

There is one chord but two trills: one for each note in the chord.  If there were only one trill, then the pianist would expect that to mean play one note with a trill and sustain the bottom note without trilling it.

The ossia in the above example shows how to cheat if you cannot play a double trill (not how it should be performed).

Test example:

```tsv
**kern
*clefG2
*M4/4
=1
2cT
2d
=2
2c 2et
2d
=3
2cT 2et
2d
=4
2cTy 2et
2d
=
*-
```

<details>
<summary> Click to view MEI conversion of above Humdrum data </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-01-16T12:29:19" version="3.15.0-dev-244e188-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m1xzrvja">
    <score xml:id="sa4w5z4">
     <scoreDef xml:id="s1n2uk90">
      <staffGrp xml:id="sf00w24">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
        <meterSig xml:id="metersig-L3F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="2" oct="4" pname="c" accid.ges="n" />
         <note xml:id="note-L6F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <trill xml:id="trill-L5F1" staff="1" startid="#note-L5F1" />
      </measure>
      <measure xml:id="measure-L7" n="2">
       <staff xml:id="staff-L7F1N1" n="1">
        <layer xml:id="layer-L7F1N1" n="1">
         <chord xml:id="chord-L8F1" dur="2">
          <note xml:id="note-L8F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L8F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L9F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <trill xml:id="trill-L8F1S2" staff="1" startid="#note-L8F1S2" />
      </measure>
      <measure xml:id="measure-L10" n="3">
       <staff xml:id="staff-L10F1N1" n="1">
        <layer xml:id="layer-L10F1N1" n="1">
         <chord xml:id="chord-L11F1" dur="2">
          <note xml:id="note-L11F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L11F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L12F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <trill xml:id="trill-L11F1S1" staff="1" startid="#note-L11F1S1" />
      </measure>
      <measure xml:id="measure-L13" n="4">
       <staff xml:id="staff-L13F1N1" n="1">
        <layer xml:id="layer-L13F1N1" n="1">
         <chord xml:id="chord-L14F1" dur="2">
          <note xml:id="note-L14F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L14F1S2" oct="4" pname="e" accid.ges="n" />
         </chord>
         <note xml:id="note-L15F1" dur="2" oct="4" pname="d" accid.ges="n" />
        </layer>
       </staff>
       <trill xml:id="trill-L14F1S1" staff="1" startid="#note-L14F1S1" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Current rendering in verovio:


Screenshot 2023-01-16 at 12 37 23 PM


The double trill in measure 3 is incorrect, but that is because the Humdrum-to-MEI conversion is incorrect, with only a single trill being converted:


```xml
       <trill xml:id="trill-L11F1S1" staff="1" startid="#note-L11F1S1" />
```

It should instead be converted as two trills, with the bottom note's trill placed below the staff automatically

```
       <trill xml:id="trill-L11F1S1" staff="1" startid="#note-L11F1S1" place="below" />
       <trill xml:id="trill-L11F1S2" staff="1" startid="#note-L11F1S2" />
```

It is best to add `@place="above"` to the top note's trill as well, so that it matches the behavior of modents being converted on a chord (with two notes).  Here are the modents from measure 3 of the top example in this thread:

```xml
       <mordent xml:id="mordent-L11F1S1" staff="1" startid="#note-L11F1S1" place="below" form="upper" />
       <mordent xml:id="mordent-L11F1S2" staff="1" startid="#note-L11F1S2" place="above" form="upper" />
```

So you could also fix that problem with this PR if you want...

**Comment by craigsapp** [comment]
> So you could also fix that problem with this PR if you want...

Are you in progress with anything extra in the PR?  If not, then I can clean up linking ornaments in chords to individual notes this weekend.

**Comment by gregchapman-dev** [comment]
I am not. Got distracted by other work. Thanks!On Jan 23, 2023, at 10:22 PM, Craig Stuart Sapp ***@***.***> wrote:﻿

So you could also fix that problem with this PR if you want...

Are you in progress with anything extra in the PR?  If not, then I can clean up linking ornaments in chords to individual notes this weekend.

**Comment by gregchapman-dev** [comment]
@craigsapp Did you make any progress here?  I can take over again if you like.

**Comment by craigsapp** [comment]
I haven't gotten to it yet, so if you want to work on it, I will let you :-)

**Comment by gregchapman-dev** [comment]
I will leave it to you for the moment; lots of other work going on.  I'll poke you for status if I have time to pick it up again.

**Comment by craigsapp** [comment]
OK, I will see about doing it in the meantime.

For the first example, I see you have implemented it attachment by `@startid` for the delayed turn:

Screenshot 2023-02-12 at 7 06 55 PM

```tsv
**kern
*M4/4
!LO:TX:b:t=not delayed
2dsSss
2e
=1
!LO:TX:b:t=delayed
2dSss
2e
==
*-
```

<details>
<summary> Click to view MEI conversion of above data </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-02-12T19:04:07" version="3.15.0-dev-7c92fc3">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mjgp1a">
    <score xml:id="s1d5nu8w">
     <scoreDef xml:id="s1l8mrfh">
      <staffGrp xml:id="sv88p2i">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="c1bsgq6b" shape="G" line="2" />
        <meterSig xml:id="metersig-L2F1" count="4" unit="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="2" oct="4" pname="d" accid.ges="n" />
         <note xml:id="note-L5F1" dur="2" oct="4" pname="e">
          <accid xml:id="accid-L5F1" accid="n" func="caution" />
         </note>
        </layer>
       </staff>
       <turn xml:id="turn-L4F1" staff="1" startid="#note-L4F1" accidupper="f" accidlower="s" form="upper" />
       <dir xml:id="dir-L3F1" place="below" staff="1" tstamp="1.000000">
        <rend xml:id="r1m5ug5v" fontstyle="normal">not delayed</rend>
       </dir>
      </measure>
      <measure xml:id="measure-L6" right="end" n="1">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L8F1" dur="2" oct="4" pname="d" accid.ges="n" />
         <note xml:id="note-L9F1" dur="2" oct="4" pname="e">
          <accid xml:id="accid-L9F1" accid="n" func="caution" />
         </note>
        </layer>
       </staff>
       <turn xml:id="turn-L8F1" staff="1" startid="#note-L8F1" accidupper="f" accidlower="s" delayed="true" form="upper" />
       <dir xml:id="dir-L7F1" place="below" staff="1" tstamp="1.000000">
        <rend xml:id="r4cigtl" fontstyle="normal">delayed</rend>
       </dir>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

The delayed

```xml
       <turn xml:id="turn-L4F1" staff="1" startid="#note-L4F1" accidupper="f" accidlower="s" form="upper" />
       <turn xml:id="turn-L8F1" staff="1" startid="#note-L8F1" accidupper="f" accidlower="s" delayed="true" form="upper" />
```

Both delayed and undelayed are attached via startid on the note.

**Comment by craigsapp** [comment]
Commit https://github.com/rism-digital/verovio/commit/fcb6fc60b817fd237abbe71e95f5d61b6d88a578 implements turns on notes in chords (only one turn on the highest note of the chord is checked, however).

There is a problem with `turn@delayed="true"` described in issue https://github.com/rism-digital/verovio/issues/3266

Trills and mordents on chord notes have yet to be implemented.
